proposals/0000-transient-storages.md

```diff
@@ -0,0 +1,336 @@
+---
+Title: transient storage host functions
+Number: 0
+Status: Proposed
+Authors:
+  - cheme
+Created: 2023-04-xx
+Category: Runtime Environment
+Requires:
+Replaces:
+--- 
+
+
+Current Implementation: https://github.com/cheme/substrate/tree/transient
+
+
+## Summary
+
+This PPPs defines new host functions to keep trace of transient data accross a block processing.
+
+It is derived and modified from initial writing https://github.com/paritytech/substrate/issues/12577, credits to Gavin Wood.
+
+It exposes two new sets of host functions, blob storage an api around bytes storage (similar to rust  `Vec<u8>`), and ordered key values storage (similar to rust `BTreeMap<Vec<u8>, Vec<u8>>` storage).
+
+Each storage structure are defined and address by a name with no limit in number except the block weight.
+Each structure can be hashed or reduced to a single merkle root to possible store it on state.
+Each structure can define if it should be send to client storage.
+
+## Motivation
+
+From https://github.com/paritytech/substrate/issues/5396: 
+"Right now we abuse storage for intra-block data such as block number, parent hash and block author as well as various housekeeping information and flags like whether we set the uncles Authorship::DidSetUncles.
+
+When initially writing, this incurs an extra trie lookup, which is slow. Instead there should be another host API, which works exactly like set_storage/get_storage but has no trie backing, so it never tries to lookup the value in the trie, nor does it write the value at the end of the block." Credits Gavin Wood.
+
+Additionally content such as events often are used in a log based manner (append only) with possibly a larger size than usual content.
+
+Hashing through host function involves passing all data at once, and is not memory efficient.
```

Comment:
From this sentence, it seems that one alternative could be to provide new host functions that allow hashing data progressively, which seems tremendously more simple than all this machinery.

If this alternative is not viable, this section really should explain why.

Reply:
This is what I propose in the paragraph "Implementation of Blob storage hashing". This part is not really strictly needed for transient storage, but it did not make sense to me to have blob (potentially big) hashing with the current api.

Maybe I should extract it in a separate PPP (progressive hashing host function).

Reply:
My point is that the "Motivation" and/or "Alternatives" section should explain what problem is being solved, but also why this specific design was chosen and not a different one. The objective is to make sure that this isn't an XY problem.

If the problem that is being solved is just that "hashing through host function isn't memory efficient", then the solution in this RFC is way overkill.

Reply:
> Similarily the ordered map solve a precise issue, and blob a slightly different one.

But again, my point is that it should be written in the PPP should issue is being solved.
It's not possible to have an opinion on whether a proposal is good if you don't know which use-cases it has and problems it solves.